### ISSUE TITLE
Needs better checking before deployment

### ISSUE THREAD

**Issue by nemesis519** [criteo-cookbooks/ms_dotnet#86]
In it's current form, the dotnet package is downloaded from MS along with updates. The package is extracted and installed along with updates (Three or four installer processes total) . During subsequent chef runs, the same events occur (extract the package, execute three or four processes) until windows realizes the packages are already installed, then it aborts.

This behaviour is ok for an underutilized server, but when running on a small instance that's under decent load, this process brings the host to an almost unresponsive state and will never complete before the next chef run. This really needs a better process to check and see if dotnot is already installed without having to extract and execute the multiple installers.

**Comment by Annih** [comment]
Hello @nemesis519 
Thank you, this is an interesting feedback, and it's clearly not the expected behavior. I'll have a look.
Could you share with me the following information to help me reproduce and debug:

1. Versions of: Windows, Chef-client, .NET
2. An extract of the Chef-client logs, showing the updated resources and executed actions
3. How do you know Chef is actually performing these actions? Did you perform a procmon, or something else?

**Comment by nemesis519** [comment]
Windows Server 2008 R2, Chef-Client 13.8.1, and .Net 4.7.1

When I manually execute a chef-client run and gets to the .Net deployment step, the MS package extractor appears on screen for 5 - 10 seconds, then three - four TrustedInstaller.exe processes start. If the server is idle, this whole process takes 30 seconds or so before the server notices the software is already installed and aborts. Then the chef-client run notes "Up-To-Date" and moves on to the next recipe. 

If the server is under load (80% cpu) from our application and a chef-client run happens, the MS package extractor appears, takes upwards of two minutes to extract, then the three - four TrustedInstaller.exe processes start.  These processes just hang and eventually run into the next chef-client run 30 minutes later.

I extracted the files out of the MS package and it includes it's the .net installer and several msu update packages. So that would explain way there are a number of TrustedInstaller processes started. 

I don't have access to the request logs at the moment, but I will try to upload them tomorrow morning. 

I think a solution to this issue would be (and I could be wrong) to setup a check to see if the software is installed or not before going through the process of extracting the package / running the installers.

Thanks for the quick response!

**Comment by Annih** [comment]
Your suggested solution is right :) but Chef "package" resource is supposed to performed this check when used properly.

When you declare a resource `package ".NET Framework"` Chef is checking that the product `.NET Framework` is present in Windows products. Some .NET installers are not registering as installed products, but as Update, and this check is not working.

So in your case, either the *package name* is incorrect, or the additionals checks I added to handle the "Update packages" are wrong. The Chef log would be very useful to determine which package is not properly idempotent.

Also please provide me the version of the `ms_dotnet` cookbook you are using, just to be sure you have all the latest patches.

**Comment by nemesis519** [comment]
ms_dotnet version 4.1.0 

I ran chef-client in debug mode and have attached the log. I edited out sensitive information. 

[chef-client-debug.log](https://github.com/criteo-cookbooks/ms_dotnet/files/2077040/chef-client-debug.log)

And below is a snippet of when it gets to the ms_dotnet. I used berks to install the ms_dotnet cookbook and then made a wrapper to call for 4.7.1.  If I'm doing something wrong on my end, feel free to correct me. This is all my recipe does - 

`include_recipe 'ms_dotnet' 

ms_dotnet_framework '4.7.1' do
  action            :install
  include_patches   true
  perform_reboot    true
  timeout			1200
end`

`[2018-06-06T15:03:11+00:00] INFO: Processing ms_dotnet_framework[4.7.1] action install (################################-dotnet471::default line 18)
[2018-06-06T15:03:11+00:00] DEBUG: Resources for generic windows_package resource enabled on node include: [Chef::Resource::WindowsPackage]
[2018-06-06T15:03:11+00:00] DEBUG: Resource for windows_package is Chef::Resource::WindowsPackage
[2018-06-06T15:03:11+00:00] DEBUG: Resources for generic ms_dotnet_reboot resource enabled on node include: [Custom resource ms_dotnet_reboot from cookbook ms_dotnet]
[2018-06-06T15:03:11+00:00] DEBUG: Resource for ms_dotnet_reboot is Custom resource ms_dotnet_reboot from cookbook ms_dotnet
[2018-06-06T15:03:11+00:00] DEBUG: Resources for generic windows_package resource enabled on node include: [Chef::Resource::WindowsPackage]
[2018-06-06T15:03:11+00:00] DEBUG: Resource for windows_package is Chef::Resource::WindowsPackage
[2018-06-06T15:03:11+00:00] INFO: Processing ms_dotnet_reboot[Reboot for ms_dotnet[4.7.1]] action nothing (C:/chef/cache/cookbooks/ms_dotnet/resources/framework.rb line 64)
[2018-06-06T15:03:11+00:00] DEBUG: Skipping ms_dotnet_reboot[Reboot for ms_dotnet[4.7.1]] due to action :nothing
[2018-06-06T15:03:11+00:00] INFO: Processing windows_package[Microsoft .NET Framework 4.7.1] action install (C:/chef/cache/cookbooks/ms_dotnet/resources/framework.rb line 83)
[2018-06-06T15:03:11+00:00] DEBUG: Providers for generic windows_package resource enabled on node include: [Chef::Provider::Package::Windows]
[2018-06-06T15:03:11+00:00] DEBUG: Provider for action install on resource windows_package[Microsoft .NET Framework 4.7.1] is Chef::Provider::Package::Windows
[2018-06-06T15:03:12+00:00] DEBUG: windows_package[Microsoft .NET Framework 4.7.1] is EXE with type 'custom'
[2018-06-06T15:03:12+00:00] DEBUG: Finding uninstall entries for Microsoft .NET Framework 4.7.1
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'QuietUninstallString'
[2018-06-06T15:03:12+00:00] DEBUG: Creating uninstall entry for [#<Win32::Registry key="HKEY_LOCAL_MACHINE">, 131353]::{92FB6C44-E685-45AD-9B20-CADF4CABA132} - 1033
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'QuietUninstallString'
[2018-06-06T15:03:12+00:00] DEBUG: Creating uninstall entry for [#<Win32::Registry key="HKEY_LOCAL_MACHINE">, 131353]::{E0C7523C-686B-3EE6-8FB1-CB4339E30EDD}
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:12+00:00] DEBUG: Registry error opening hive '#<Win32::Registry::PredefinedKey:0x0000000004d67898>' :: 131097: The system cannot find the file specified.
[2018-06-06T15:03:12+00:00] DEBUG: windows_package[Microsoft .NET Framework 4.7.1] checking package version
[2018-06-06T15:03:12+00:00] INFO: Processing remote_file[C:\chef\cache\package\NDP471-KB4033342-x86-x64-AllOS-ENU.exe] action create (dynamically defined)
[2018-06-06T15:03:12+00:00] DEBUG: remote_file[C:\chef\cache\package\NDP471-KB4033342-x86-x64-AllOS-ENU.exe] checksumming file at C:\chef\cache\package\NDP471-KB4033342-x86-x64-AllOS-ENU.exe.
[2018-06-06T15:03:12+00:00] DEBUG: remote_file[C:\chef\cache\package\NDP471-KB4033342-x86-x64-AllOS-ENU.exe] checking for changes
[2018-06-06T15:03:12+00:00] DEBUG: remote_file[C:\chef\cache\package\NDP471-KB4033342-x86-x64-AllOS-ENU.exe] checksum matches target checksum (63dc850df091f3f137b5d4392f47917f847f8926dc8af1da9bfba6422e495805) - not updating
[2018-06-06T15:03:12+00:00] DEBUG: windows_package[Microsoft .NET Framework 4.7.1] fetched source file to C:\chef\cache\package\NDP471-KB4033342-x86-x64-AllOS-ENU.exe
[2018-06-06T15:03:12+00:00] DEBUG: windows_package[Microsoft .NET Framework 4.7.1] checking package version
[2018-06-06T15:03:12+00:00] DEBUG: windows_package[Microsoft .NET Framework 4.7.1] Microsoft .NET Framework 4.7.1 ["4.7.02558"] already installed
[2018-06-06T15:03:12+00:00] DEBUG: windows_package[Microsoft .NET Framework 4.7.1] is already installed - nothing to do
[2018-06-06T15:03:12+00:00] INFO: Processing windows_package[Update for Microsoft Windows (KB4054856)] action install (C:/chef/cache/cookbooks/ms_dotnet/resources/framework.rb line 83)
[2018-06-06T15:03:13+00:00] DEBUG: Providers for generic windows_package resource enabled on node include: [Chef::Provider::Package::Windows]
[2018-06-06T15:03:13+00:00] DEBUG: Provider for action install on resource windows_package[Update for Microsoft Windows (KB4054856)] is Chef::Provider::Package::Windows
[2018-06-06T15:03:13+00:00] DEBUG: windows_package[Update for Microsoft Windows (KB4054856)] is EXE with type 'custom'
[2018-06-06T15:03:13+00:00] DEBUG: Finding uninstall entries for Update for Microsoft Windows (KB4054856)
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Failure to read property 'DisplayName'
[2018-06-06T15:03:13+00:00] DEBUG: Registry error opening hive '#<Win32::Registry::PredefinedKey:0x0000000004d67898>' :: 131097: The system cannot find the file specified.
[2018-06-06T15:03:13+00:00] DEBUG: windows_package[Update for Microsoft Windows (KB4054856)] checking package version
[2018-06-06T15:03:13+00:00] INFO: Processing remote_file[C:\chef\cache\package\NDP471-KB4054856-x86-x64-AllOS.exe] action create (dynamically defined)
[2018-06-06T15:03:13+00:00] DEBUG: remote_file[C:\chef\cache\package\NDP471-KB4054856-x86-x64-AllOS.exe] checksumming file at C:\chef\cache\package\NDP471-KB4054856-x86-x64-AllOS.exe.
[2018-06-06T15:03:13+00:00] DEBUG: remote_file[C:\chef\cache\package\NDP471-KB4054856-x86-x64-AllOS.exe] checking for changes
[2018-06-06T15:03:13+00:00] DEBUG: remote_file[C:\chef\cache\package\NDP471-KB4054856-x86-x64-AllOS.exe] checksum matches target checksum (f98d12d84c803699d149a254ada3dd5dc0698307638e1baae209cc6f3a729e29) - not updating
[2018-06-06T15:03:13+00:00] DEBUG: windows_package[Update for Microsoft Windows (KB4054856)] fetched source file to C:\chef\cache\package\NDP471-KB4054856-x86-x64-AllOS.exe
[2018-06-06T15:03:13+00:00] DEBUG: windows_package[Update for Microsoft Windows (KB4054856)] checking package version
[2018-06-06T15:03:13+00:00] DEBUG: windows_package[Update for Microsoft Windows (KB4054856)] Update for Microsoft Windows (KB4054856) not installed, installing latest
[2018-06-06T15:03:13+00:00] DEBUG: windows_package[Update for Microsoft Windows (KB4054856)] installing  package 'c:\chef\cache\package\ndp471-kb4054856-x86-x64-allos.exe'
[2018-06-06T15:03:47+00:00] INFO: windows_package[Update for Microsoft Windows (KB4054856)] installed Update for Microsoft Windows (KB4054856) at latest
[2018-06-06T15:03:47+00:00] INFO: windows_package[Update for Microsoft Windows (KB4054856)] sending reboot_if_pending action to ms_dotnet_reboot[Reboot for ms_dotnet[4.7.1]] (immediate)
[2018-06-06T15:03:47+00:00] INFO: Processing ms_dotnet_reboot[Reboot for ms_dotnet[4.7.1]] action reboot_if_pending (C:/chef/cache/cookbooks/ms_dotnet/resources/framework.rb line 64)
[2018-06-06T15:03:47+00:00] DEBUG: Resources for generic reboot resource enabled on node include: [Chef::Resource::Reboot]
[2018-06-06T15:03:47+00:00] DEBUG: Resource for reboot is Chef::Resource::Reboot
[2018-06-06T15:03:47+00:00] INFO: Processing reboot[Reboot for windows_package[Update for Microsoft Windows (KB4054856)]] action reboot_now (C:/chef/cache/cookbooks/ms_dotnet/resources/reboot.rb line 28)
[2018-06-06T15:03:47+00:00] DEBUG: Skipping reboot[Reboot for windows_package[Update for Microsoft Windows (KB4054856)]] due to only_if ruby block`

**Comment by Annih** [comment]
Ok I got it:

- According to your snippet, the faulty package is the `Update for Microsoft Windows (KB4054856)`, which generates many logs and is not detected as "installed". Your first investigation was already pointing to an update package :)
- [KB4054856](https://support.microsoft.com/help/4054856/) is a meta KB package including `KB4054852` for WS2008R2, `KB405853` for WS2012, `KB405854` for WS2012R2 and `KB405855` for WS2016. 
- As I told you, for packages installed as an Update I implemented an additional mechanism. I used it with the KB number listed above to "guard" the package setup, and I put a "meta" name for the package `Update for Microsoft Windows (KB4054856)`.

But the problem is that I misread the [KB4054856 documentation](https://support.microsoft.com/help/4054856/), on Windows Server 2008R2 it is installed as a product `Update for Microsoft.NET Framework 4.7.1 (KB4054852)`, so neither the product name check and the additional KB check worked.

I'll provide a  patch soon :)
